### DISCUSSION TITLE
Having an ILogger shut down when my ASP.NET Core app shuts down

### DISCUSSION THREAD

**Discussion by DavidThielen** [dotnet/aspnetcore#52490] [Q&A]
Hi all;

This is in a Blazor server app. I have an ILoggerProvider that I created. It buffers log output and writes it (to an Azure BLOB) every 2 minutes. When the server shuts down, I need to have it flush the buffered output and write it to the BLOB.

Is there a way to do this? I implemented `Dispose()` but it appears to not be called on a shutdown. Do I need to register my ILoggerProvider  in some special way?

And if there is no way in the ILogger framework to do this, is there a way to have the App Server (I use Azure) call a method in my code on shutdown? If so, I can then directly call the logger to flush.

thanks - dave

**Comment by En3Tho** [comment]
I think as an immediate workaround you can implement a hosted service that will release resources of your provider.

**Reply by DavidThielen** [reply]
I'm sorry but that's a terrible answer. That turns a lightweight class into a service and all the overhead that goes with it. And then when I scale out my web app, I then need to scale out these service instances.

At worst all I need is an event that is called when the app server is shutting down. In that I make a simple call to the logger objects.

**Reply by En3Tho** [reply]
I don't get the scaling point tbh. Hosted service is just a class that reacts to lifetime events (and that's what you need).

But I can understand the point of logic being split between classes simply to call a dispose once. That's why I called it an immediate workaround.

**Comment by davidfowl** [comment]
It should be called on shutdown, how did you add the provider to the DI container?

**Reply by DavidThielen** [reply]
All I do is the following (if there's more I should do, please let me know):

The provider:
```cs
	[ProviderAlias("Blob")]
	public class BlobLoggerProvider : LoggerProviderBase
	{
        // ...
        }

public abstract class LoggerProviderBase : ILoggerProvider
{
// ...
	public void Dispose()
	{
		Enqueue("***** Dispose called *****");
		_queueWriter?.FlushQueue();
		SettingsChangeToken?.Dispose();
		SettingsChangeToken = null;
		_timerQueueWriter?.Dispose();
		_timerQueueWriter = null;
		GC.SuppressFinalize(this);
	}
}
```

appSettings.json
```json
   "Logging": {
       "LogLevel": {
           "Default": "Information",
           "LouisHowe": "Debug",
           "ThirdPartyServices": "Information",
           "CommonUtilities": "Information",
           "Microsoft": "Warning"
       },
       "Blob": {
           "Path": "logs/{yyyy-MM-dd}.txt",
           "Interval": "Day",
           "AzureStorage": "UseDevelopmentStorage=true",
           "FlushInterval": 1
       },
       "File": {
           "Path": "C:/temp/log/{yyyy-MM-dd}.txt",
           "Interval": "day",
           "FlushInterval": 1
       },
```

And how should it call it on shutdown?

**Reply by DavidThielen** [reply]
BTW - off-topic but want to say that SignalR is beautifully designed.

**Reply by davidfowl** [reply]
How do you add it to the container?

**Reply by DavidThielen** [reply]
I don't add it to the container. I just searched and couldn't find anything about adding an `ILoggerProvider` to a container. Can you point me to a page that tells me how to do that?

I'm guessing that explains the issue - not in a container, then no way for Blazor to know to Dispose it on shutdown.

TIA

**Reply by davidfowl** [reply]
What does the wire up of your BlobLoggerProvider look like?

**Reply by DavidThielen** [reply]
First off, thanks for continuing to help on this. I add it to the system as follows. In Program.cs, in `Main()` I have:

```cs
builder.Logging.AddAzureBlobLogger();
```

And that method is:

```cs
public static ILoggingBuilder AddAzureBlobLogger(this ILoggingBuilder builder)
{
	builder.AddConfiguration();

	builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, BlobLoggerProvider>());
	builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<BlobLoggerOptions>, BlobLoggerOptionsSetup>());
	builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IOptionsChangeTokenSource<BlobLoggerOptions>, LoggerProviderOptionsChangeTokenSource<BlobLoggerOptions, BlobLoggerProvider>>());

	return builder;
}
```

Almost all of what I did, from the `[ProviderAlias("Blob")]` to this code just shown I came up with by copying from the Console ILoggerProvider. There wasn't any documentation I could find on writing one. So if I missed a required step, it's because I didn't see it for the Console provider.

thanks - dave
